common/elasticsearch/client_v6.go

```diff
@@ -0,0 +1,664 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package elasticsearch
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"math"
+	"strconv"
+	"time"
+
+	"github.com/olivere/elastic"
+
+	workflow "github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/metrics"
+	p "github.com/uber/cadence/common/persistence"
+	"github.com/uber/cadence/common/service/config"
+	"github.com/uber/cadence/common/types/mapper/thrift"
+)
+
+var _ GenericClient = (*elasticV6)(nil)
+var _ GenericBulkProcessor = (*v6BulkProcessor)(nil)
+
+type (
+	// elasticV6 implements Client
+	elasticV6 struct {
+		client     *elastic.Client
+		config     *config.VisibilityConfig
+		logger     log.Logger
+		serializer p.PayloadSerializer
+	}
+
+	// searchParametersV6 holds all required and optional parameters for executing a search
+	searchParametersV6 struct {
+		Index       string
+		Query       elastic.Query
+		From        int
+		PageSize    int
+		Sorter      []elastic.Sorter
+		SearchAfter []interface{}
+	}
+
+	// bulkProcessorParametersV6 holds all required and optional parameters for executing bulk service
+	bulkProcessorParametersV6 struct {
+		Name          string
+		NumOfWorkers  int
+		BulkActions   int
+		BulkSize      int
+		FlushInterval time.Duration
+		Backoff       elastic.Backoff
+		BeforeFunc    elastic.BulkBeforeFunc
+		AfterFunc     elastic.BulkAfterFunc
+	}
+)
+
+var (
+	oneMilliSecondInNano = int64(1000)
+)
+
+// NewWrapperClient returns a new implementation of Client
+func newV6Client(
+	connectConfig *config.ElasticSearchConfig,
+	visibilityConfig *config.VisibilityConfig,
+	logger log.Logger,
+) (GenericClient, error) {
```

Comment:
Return struct (pointer) instead. And add type assertion.

Reply:
No I think we should return interface. Same as the `ExponentialBackoff`. If returning struct, the struct needs to be public, otherwise you will get compiler warning. I don't see it's necessary to return the v6 struct here.